Fix tests to allow them to run in parallel without conflict

samdoran opened the pull request:
Ensure a custom copy of the data directory is used for each test rather than having multiple tests share the same directory. This means the default tox environment no longer needs to have two separate `pytest` commands.

- Change the tests so they do not create artifacts in the source tree and remove corresponding ignore entries.
- Remove custom `pytest` marks and define `testpaths` in `pytest.ini`.
- Consolidate test fixtures and rename "data" to "projects"
- Various improvements to playbooks and inventory used it tests
- Disable Ansible devel warning in containerized tests
- Change `yamllint` settings to be a bit more sane
- Add a custom marker that will generate tests for each supported container runtime

samdoran committed:
Make yamllint settings a bit more sane

samdoran committed:
Remove custom serial markers

Run all tests together.
Use pytest.ini to configure default test path.

samdoran committed:
Create a separate copy of the data for each test

Remove ignore entries since tests should not be writing data to the source tree

samdoran committed:
Update tests to use separate data dirs and Path objects

samdoran committed:
Reorginize test fixtures

Add a pytest fixture for easily accessing the project path.

samdoran committed:
Remove inventory files defining explicit localhost.

samdoran committed:
Use project fixture and remove custom tmp file management

samdoran committed:
Remove printenv_example fixture

Now that tests use separate directories, cleaning out env/ is unnecessary.

samdoran committed:
Use unique container names to avoid conflicts

AlanCoding commented:
This is great to see. I know it's a lot of work. I had tried to work at the edges of some of these issues before and got really overwhelmed.

I think I introduced the `test/data` folder. Everything you're doing here looks right, like, so obvious there is no question about correctness, but just difficulty in actually getting it working. Let me know if I can help to get it past the finish line.

(and really, it looks like the only CI failures are the linter failures which are already marked in the diff, plus the regular Zuul flake)

AlanCoding commented on test/conftest.py:
```diff
@@ -25,6 +28,15 @@ def skipif_pre_ansible28(is_pre_ansible28):
+def test_data_dir(tmp_path):
+    source = Path(__file__).parent / 'fixtures' / 'data'
+    dest = tmp_path / 'data'
+    shutil.copytree(source, dest)
```
clever.

This is function scope, and copies the entire data directory? As long as the user clears old artifacts before running, that's 14 dirs / 14 files. All of these tend to be very small in the scheme of things, so this feels like the right kind of balance to me. :+1:

samdoran replied:
The more I look at this, the more it seems like it was a collection of project directories and only one is needed per test? I don't think the whole thing needs to be duplicated for each test, but that's what I did before I started to realize what these directories were. It's a bit blunt, but seems to adequately isolate the tests.

I'm open to better ways to do this rather than copying the whole directory structure around.

> This is function scope, and copies the entire data directory?

Yes.

> As long as the user clears old artifacts before running

Hopefully there are _no_ artifacts in these directories. That was the goal. I removed the `.gitignore` entries that would hide errant tests depositing artifacts in the source tree.

AlanCoding reviewed (approved):
looks like you might only need to fix the linter issues before this can be merged

samdoran committed:
Remove unused imports

samdoran commented:
> looks like you might only need to fix the linter issues before this can be merged

I wish. 😞

It looks like the CI environment is different than my local environment in some way. `ansible-runner-tox-ansible-base` and `ansible-runner-tox-ansible29` are failing becasue they cannot find the playbooks inside the temporary directory. I'm unable to reproduce it locally (on Linux on macOS) so I'm not sure what's different in the CI environment.

AlanCoding commented on test/integration/containerized/test_cli_containerized.py:
```diff
@@ -12,20 +12,17 @@ def skip_if_no_podman(container_runtime_installed):
 
-@pytest.mark.serial
 def test_module_run(cli, skip_if_no_podman):
     r = cli(['run', '-m', 'ping', '--hosts', 'localhost', os.path.join(HERE, 'priv_data')])
```
don't you need to move the `priv_data` folder into the `test/data` folder? Otherwise, this isn't ready to be ran in parallel.

samdoran replied:
Yup, I missed that one.

samdoran committed:
Cleanup new changes from devel

samdoran committed:
Run all tests on supported runtimes rather than just podman

Working on refactoring this into something common and reusable.

samdoran committed:
Disable Anisble devel warning in test fixture projects

samdoran committed:
Improve test inventory

Explicitly define hosts that use local connection and the expected
Python interpreter.

Setting connection: local in the playbook without specifying the interpreter
causes interpreter discovery to run, which could result in the incorrect
interpreter being selected.

samdoran committed:
Move fixture data used common directory

samdoran committed:
Use existing facts for determing containerization state

samdoran committed:
Use existing sleep project in test

We already have a playbook that sleeps for a bit, so just reuse that.

samdoran committed:
Consolidate and move project fixtures

Now that I realize a bit more what these are, I consolidated all project
fixtures into a single directory, renamed it to "projects", and removed use
of the old fixture.

That should be the last test that was leaving artifacts in the source tree.

samdoran committed:
Auto generate test parameters for testing all supported runtimes

This happens based on marking tests with a custom mark. That may be a bit
too magic, but it does work quite well.

samdoran committed:
Rename fixture to better reflect its content

samdoran committed:
Move fixtures only used by integration tests into that directory

samdoran committed:
Set container env var for consistent test results
